File: libs/authz-cedar/src/engine.rs
//! Top-level engine orchestrating policy evaluation, audit emission
//! and entity hydration.
//!
//! Composition:
//!
//! ```text
//!   AuthzEngine
//!     ├── PolicyStore        (Arc<RwLock<PolicySet>> + bundled schema)
//!     ├── AuditSinkHandle    (Arc<dyn AuthzAuditSink>)
//!     └── (entities are caller-supplied per request)
//! ```
//!
//! `authorize` is the canonical entry point. The Axum extractor
//! (`crate::axum::AuthzGuard`) and the test-suite both exercise this
//! method; the underlying `PolicyStore::is_authorized` is exposed for
//! callers that need raw Cedar `Response`s.

use std::sync::Arc;

use cedar_policy::{Context, Decision, Entities, EntityUid, Request};

use crate::{
    PolicyStore, PolicyStoreError,
    audit::{AuditSinkHandle, AuthzAuditEvent, NoopAuditSink},
};

/// Combined handle: policy store + audit sink.
#[derive(Clone)]
pub struct AuthzEngine {
    store: PolicyStore,
    audit: AuditSinkHandle,
}

impl AuthzEngine {
    /// Build an engine from a [`PolicyStore`] and an audit sink.
    pub fn new(store: PolicyStore, audit: AuditSinkHandle) -> Self {
        Self { store, audit }
    }

    /// Convenience constructor with a [`NoopAuditSink`] — the canonical
    /// choice for unit tests.
    pub fn with_noop_audit(store: PolicyStore) -> Self {
        Self::new(store, Arc::new(NoopAuditSink))
    }

    /// Underlying [`PolicyStore`] handle.
    pub fn store(&self) -> &PolicyStore {
        &self.store
    }

    /// Audit sink handle (cloning is cheap).
    pub fn audit(&self) -> AuditSinkHandle {
        Arc::clone(&self.audit)
    }

    /// Canonical authorization API.
    ///
    /// Returns the raw [`Decision`] together with the diagnostic
    /// information Cedar surfaces (matching policy ids and any
    /// evaluation errors). The audit event is **emitted before the
    /// method returns** but on a detached task, so it never stalls
    /// the request hot path. Callers that need the audit emission to
    /// be synchronous must call the sink directly.
    pub async fn authorize(
        &self,
        principal: EntityUid,
        action: EntityUid,
        resource: EntityUid,
        context: Context,
        entities: &Entities,
    ) -> Result<AuthorizeOutcome, PolicyStoreError> {
        let request = Request::new(
            principal.clone(),
            action.clone(),
            resource.clone(),
            context,
            Some(&self.store.schema()),
        )
        .map_err(|e| PolicyStoreError::Validation(e.to_string()))?;
        let response = self.store.is_authorized(&request, entities).await;

        let decision = response.decision();
        let policy_ids: Vec<String> = response
            .diagnostics()
            .reason()
            .map(|p| p.to_string())
            .collect();
        let diagnostics: Vec<String> = response
            .diagnostics()
            .errors()
            .map(|e| e.to_string())
            .collect();

        let event = AuthzAuditEvent {
            timestamp: chrono::Utc::now(),
            principal: principal.to_string(),
            action: action.to_string(),
            resource: resource.to_string(),
            decision: match decision {
                Decision::Allow => "allow".into(),
                Decision::Deny => "deny".into(),
            },
            tenant: None,
            policy_ids: policy_ids.clone(),
            diagnostics: diagnostics.clone(),
        };
        let sink = Arc::clone(&self.audit);
        tokio::spawn(async move { sink.emit(event).await });

        Ok(AuthorizeOutcome {
            decision,
            policy_ids,
            diagnostics,
        })
    }
}

/// Result returned by [`AuthzEngine::authorize`].
#[derive(Debug, Clone)]
pub struct AuthorizeOutcome {
    pub decision: Decision,
    pub policy_ids: Vec<String>,
    pub diagnostics: Vec<String>,
}

impl AuthorizeOutcome {
    pub fn is_allow(&self) -> bool {
        matches!(self.decision, Decision::Allow)
    }
}

File: libs/authz-cedar-go/engine.go
package cedarauthz

import (
	"context"
	"time"

	cedar "github.com/cedar-policy/cedar-go"
)

// AuthzEngine orchestrates policy evaluation, audit emission, and
// entity hydration. Composition mirrors the Rust impl:
//
//	AuthzEngine
//	  ├── *PolicyStore   (policy set + bundled schema)
//	  └── AuthzAuditSink (audit emission, fire-and-forget)
//
// Authorize is the canonical entry point. Callers that need raw
// diagnostics without audit emission can call PolicyStore.IsAuthorized
// directly.
type AuthzEngine struct {
	store *PolicyStore
	audit AuthzAuditSink
}

// NewEngine builds an engine from a [*PolicyStore] and an audit sink.
// Pass [NoopAuditSink]{} for tests.
func NewEngine(store *PolicyStore, audit AuthzAuditSink) *AuthzEngine {
	if audit == nil {
		audit = NoopAuditSink{}
	}
	return &AuthzEngine{store: store, audit: audit}
}

// NewEngineNoopAudit is a convenience constructor matching the Rust
// `with_noop_audit` helper.
func NewEngineNoopAudit(store *PolicyStore) *AuthzEngine {
	return NewEngine(store, NoopAuditSink{})
}

// Store returns the underlying policy store handle.
func (e *AuthzEngine) Store() *PolicyStore { return e.store }

// Audit returns the configured sink.
func (e *AuthzEngine) Audit() AuthzAuditSink { return e.audit }

// AuthorizeOutcome is the result returned by [AuthzEngine.Authorize].
type AuthorizeOutcome struct {
	Decision    cedar.Decision
	PolicyIDs   []string
	Diagnostics []string
}

// IsAllow reports whether the decision was Allow.
func (o *AuthorizeOutcome) IsAllow() bool { return o.Decision == cedar.Allow }

// Authorize evaluates a Cedar request and emits an audit event in the
// background. Audit emission runs in a goroutine so a slow sink can't
// stall the request hot path — matches the Rust `tokio::spawn` pattern.
//
// Callers that need synchronous audit must call the sink directly.
func (e *AuthzEngine) Authorize(
	ctx context.Context,
	principal cedar.EntityUID,
	action cedar.EntityUID,
	resource cedar.EntityUID,
	context_ cedar.Record,
	entities cedar.EntityGetter,
) (*AuthorizeOutcome, error) {
	req := cedar.Request{
		Principal: principal,
		Action:    action,
		Resource:  resource,
		Context:   context_,
	}
	decision, diag := e.store.IsAuthorized(entities, req)

	policyIDs := make([]string, 0, len(diag.Reasons))
	for _, r := range diag.Reasons {
		policyIDs = append(policyIDs, string(r.PolicyID))
	}
	diagnostics := make([]string, 0, len(diag.Errors))
	for _, d := range diag.Errors {
		diagnostics = append(diagnostics, d.String())
	}

	decisionStr := "deny"
	if decision == cedar.Allow {
		decisionStr = "allow"
	}

	event := AuthzAuditEvent{
		Timestamp:   time.Now().UTC(),
		Principal:   principal.String(),
		Action:      action.String(),
		Resource:    resource.String(),
		Decision:    decisionStr,
		PolicyIDs:   append([]string(nil), policyIDs...),
		Diagnostics: append([]string(nil), diagnostics...),
	}
	go e.audit.Emit(detachContext(ctx), event)

	return &AuthorizeOutcome{
		Decision:    decision,
		PolicyIDs:   policyIDs,
		Diagnostics: diagnostics,
	}, nil
}

// detachContext returns a new context that carries cancellation/values
// from `parent` for cancellation purposes only. The Rust impl uses
// `tokio::spawn` which doesn't propagate the request span; we mirror
// that by giving the audit goroutine a fresh background context so
// request cancellation doesn't cancel audit emission mid-write.
func detachContext(_ context.Context) context.Context {
	return context.Background()
}
